### PR TITLE
Initial version of automation framework.

### PR DESCRIPTION
**Pull request by michael-berlin**
@alainjobart 

Ammar already reviewed https://github.com/youtube/vitess/pull/669 and I made several changes. This extra PR contains the changes squashed into the original commit i.e. perfect if you haven't looked at it yet. I will keep the other PR to push the changes as extra commits to make it easier for Ammar to go through only the changes. Once you both are fine with the general structure, I would like to commit this and make smaller iterations from there. So let's agree on the basics of the automation framework for now.

Suggest order of files to review:
- automation.proto
- scheduler_test.go
- resharding_task_test.go
- scheduler.go
- resharding_task.go
- rest of the files

**Original commit message:**

The automation framework allows to automate cluster operations which
require a series of manual steps e.g. resharding.

A Cluster Operation has a list of task containers which are processed
sequentially. Each task container can contain one or more tasks which
will be executed in parallel.

Here's an example of a cluster operation with two task containers. The
second task container has two tasks:
- step 1
- step 2a | step 2b

If the task container contains one task, the task can emit new task
containers which will be inserted after the current task container. This
mechanism is used to fully expand Cluster Operations by special tasks
which emit new task containers e.g. "ReshardingTask".

This patchset implements the minimal steps to automate "resharding". It
lacks extensive testing of error corner cases which will be added in
later commits.

### CODE REVIEW

File: go/vt/automation/resharding_task.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+	return firstKeyspace, shards, nil
+}
+
+func selectAnyTabletFromShardByType(shard string, tabletType string) string {
+	return ""
+}
+
+func (t *ReshardingTask) run(parameters map[string]string) ([]*pb.TaskContainer, string, error) {
+	sourceShardsWithKeyspace := strings.Split(parameters["source_shard_list"], ",")
```
Comment (alainjobart) [inline go/vt/automation/resharding_task.go]: These parameters are a bit confusing to me, I'd rather have (with examples):
keyspace = ruser
source_shard_list = 10-20
dest_shard_list = 10-18,18-20

File: go/vt/automation/vtctl_task.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+	pb "github.com/youtube/vitess/go/vt/proto/automation"
+)
+
+// VtctlTask executes known vtctl commands. Output is preserved.
+type VtctlTask struct {
+}
+
+func (t *VtctlTask) run(parameters map[string]string) ([]*pb.TaskContainer, string, error) {
+	// TODO(mberlin): Use the vtctl code directly.
```
Comment (alainjobart) [inline go/vt/automation/vtctl_task.go]: Yeah this is not gonna work as is, vtctl would need a lot of extra command line parameters to work. You can use vtctlclient code, or link with vtctl library (the scheduler will end up being in vtctld anyway).

File: go/vt/automation/vtworker_task.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	pb "github.com/youtube/vitess/go/vt/proto/automation"
+)
+
+// VtworkerTask triggers an action on a running vtworker process using the HTTP server and waits until the process has finished.
+// TODO(mberlin): Add an RPC interface at vtworker and use that instead.
```
Comment (alainjobart) [inline go/vt/automation/vtworker_task.go]: We talked about this, should be a pool of vtworkers that pick up jobs, I think.

File: go/vt/automation/resharding_task.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	pb "github.com/youtube/vitess/go/vt/proto/automation"
+	"github.com/youtube/vitess/go/vt/topo"
+)
+
+// ReshardingTask is a cluster operation which allows to increase the number of shards.
```
Comment (alainjobart) [inline go/vt/automation/resharding_task.go]: merges should also be supported eventually.

File: go/vt/automation/resharding_task.go (hunk 1 above)
Comment (michael-berlin) [inline go/vt/automation/resharding_task.go]: DONE

File: go/vt/automation/vtctl_task.go (hunk 2 above)
Comment (michael-berlin) [inline go/vt/automation/vtctl_task.go]: ACK

Efforts around this will be part of a different PR.

File: go/vt/automation/vtworker_task.go (hunk 3 above)
Comment (michael-berlin) [inline go/vt/automation/vtworker_task.go]: ACK

I updated the TODO.

File: go/vt/automation/resharding_task.go (hunk 4 above)
Comment (michael-berlin) [inline go/vt/automation/resharding_task.go]: ACK

Ammar suggested to rename it to HorizontalReshardingTask. When we support merges, there will be another task.

**Comment by michael-berlin** [comment]
Deleting this pull request because all changes were made to the original PR: https://github.com/youtube/vitess/pull/669
